src/distilabel/tasks/text_generation/evol_instruct.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2023-present, Argilla, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+from dataclasses import dataclass
+from typing import Any, Dict, List, Literal, Optional, get_args
+
+from distilabel.tasks.base import get_template
+from distilabel.tasks.prompt import Prompt
+from distilabel.tasks.text_generation.base import TextGenerationTask
+
+_EVOL_INSTRUCT_TEMPLATE = get_template("evol-instruct.jinja2")
+
+
+EvolutionMethod = Literal[
+    "breadth", "constraints", "deepen", "concretizing", "reasoning"
+]
+
+
+@dataclass
+class EvolInstructTask(TextGenerationTask):
+    """A `TextGenerationTask` following the `EvolInstruct` specification for building the prompts.
+
+    Args:
+        system_prompt (str, optional): the system prompt to be used. Not defined for this task.
+
+    Notes:
+        - The paper mentions a parameter *M* for the number of evolutions. This behavior can be achieved dealing
```

Comment:
can't we add this as argument to the task itself?

Reply:
Similar to the mention below, I would say the value of this task is hidden in the automated evolution of the dataset.

Reply:
Reviewing the code currently we cannot do what the parameter M implies. For example in the OpenAI implementation, it already has a `num_generations` argument which calls the API that amount of times, but currently, modifying this argument we don't call the task `generate_prompt` again. We could review the implementation, but I wanted to keep it as is for the moment

Reply:
Yes I see. I'm not too familiar with the API yet but implementing this would be cumbersome.

Reply:
I think we can emphasize the need for multi-evolutions a bit more clearly in the docs instead @plaguss

Reply:
Rewritten at the docstrings for the task.